src/util/include/util/SmallVector.h

```diff
@@ -0,0 +1,115 @@
+
+    T& operator[](size_type pos) { return nb <= N ? dataArray[pos] : dataVector[pos]; }
+    const T& operator[](size_type pos) const { return nb <= N ? dataArray[pos] : dataVector[pos]; }
+    T& front() { return nb <= N ? dataArray.front() : dataVector.front(); }
+    const T& front() const { return nb <= N ? dataArray.front() : dataVector.front(); }
+    T& back() { return nb <= N ? dataArray[nb - 1] : dataVector.back(); }
+    const T& back() const { return nb <= N ? dataArray[nb - 1] : dataVector.back(); }
+
+    T* data() { return nb <= N ? dataArray.data() : dataVector.data(); }
+    const T* data() const { return nb <= N ? dataArray.data() : dataVector.data(); }
+
+    using const_iterator = const T*;
+    using iterator = T*;
+
+    iterator begin() { return nb <= N ? dataArray.data() : dataVector.data(); }
+    const_iterator begin() const { return nb <= N ? dataArray.data() : dataVector.data(); }
+    iterator end() { return nb <= N ? dataArray.data() + nb : dataVector.data() + nb; }
+    const_iterator end() const { return nb <= N ? dataArray.data() + nb : dataVector.data() + nb; }
+
+    void push_back(const T& value) {
+        if (nb < N) {
+            dataArray[nb] = value;
+        } else if (nb == N) {
+            dataVector.reserve(N + 1);
+            std::move(dataArray.begin(), dataArray.end(), std::back_inserter(dataVector));
+            dataVector.push_back(value);
+        } else {
+            dataVector.push_back(value);
+        }
+        ++nb;
+    }
+    void push_back(T&& value) {
+        if (nb < N) {
+            dataArray[nb] = std::move(value);
+        } else if (nb == N) {
+            dataVector.reserve(N + 1);
+            std::move(dataArray.begin(), dataArray.end(), std::back_inserter(dataVector));
+            dataVector.push_back(std::move(value));
+        } else {
+            dataVector.push_back(std::move(value));
+        }
+        ++nb;
+    }
+    template <class... Args>
+    void emplace_back(Args&&... args) {
+        if (nb < N) {
+            dataArray[nb] = T(std::forward<Args>(args)...);
+        } else if (nb == N) {
+            dataVector.reserve(N + 1);
+            std::move(dataArray.begin(), dataArray.end(), std::back_inserter(dataVector));
+            dataVector.emplace_back(std::forward<Args>(args)...);
+        } else {
+            dataVector.emplace_back(std::forward<Args>(args)...);
+        }
+        ++nb;
+    }
+    void pop_back() {
+        if (nb > N) {
+            if (nb == N + 1) {
+                std::move(dataVector.begin(), std::prev(dataVector.end()), dataArray.begin());
+                dataVector.clear();
+            } else {
+                dataVector.pop_back();
+            }
+        }
+        --nb;
+    }
+    void clear() {
+        nb = 0;
+        /**
+         * Reset capacity. If clear() is called a lot and more than N objects are usually pushed, don't use this class!
+         */
+        dataVector = {};
```

Comment:
In the current state, since the invariant that all data is located at the stack is true, when nb<=N, you could just ignore dataVector, or also use `dataVector.clear()`. Just make sure you make the same decision as in `pop_back()`

Reply:
done